### DISCUSSION TITLE
Can I just avoid using the content builder?

### DISCUSSION THREAD

**Discussion by SpanishCrusader** [MonoGame/MonoGame#8597] [Content]
Hello all, hope everything is goind well for you guys.
I've been using monogame for a few years now, and I have an issue with the content builder.

I'm a screen reader user (NVDA specificly) and as far I can remember, the content builder GUI isn't very accessible by my screen reader.
So what I had to do in order to add content to my game is basically edit the Content file manually, which is a very tedious job as you get further into the development.

This issue between the content builder GUI and my screen reader has been present since I started using monogame in 2019, so I believe this issue is something I'll always have to deal with.

My question is, can I just ignore the content builder and load assets manually from a file stream without having consequences in the future?

I want to start loading all the game's content from file streams to avoid the content builder, but I don't want to regret it later.

Example: Turns out my game does well on Steam, so I decide to make a port for future consoles but oops, I have to switch to the content builder since there is no way to load content directly from a file stream.

Thanks you so much for your help!

**Comment by squarebananas** [comment]
Not all assets have a `FromStream` method currently such as `SpriteFont` and `Model`. 

However as the editor is really just a visual option to edit the `.content` files, it is possible to edit these files in a text editor instead. All pipeline build features are still available this way. It's not ideal, but you can use the editor once to produce a `.content` file for each asset type and then use that file as a template to copy/paste from.

More importantly though, the editor should be updated to be more accessibility friendly if there are issues. Someone on Discord who is severely sight impaired has recently mentioned issues with the editor as well.

Trying it myself on Windows I can see some of the issues:
- The large text mode doesn't effect all text.
- Some readable areas don't work with screen readers at all.
- Other areas work with the Windows built-in Narrator but not with built-in `Magnifier`, and vice versa.
- When clicking an asset, Magnifier reads it's name and Narrator gives an essay.

If it is possible to list any other issues they could probably be fixed.

![Screenshot 2024-12-07 105856](https://github.com/user-attachments/assets/9930ded8-9fdc-42aa-814e-34317ce3493a)

> [!NOTE]  
> Videos will probably start muted.

https://github.com/user-attachments/assets/0aab527e-da9d-4057-9005-f66aa1cad1f1

https://github.com/user-attachments/assets/28c71de6-7c2d-4691-963d-dbb9f5ed9ba1

**Comment by dellis1972** [comment]
I have opened an issue for this https://github.com/MonoGame/MonoGame/issues/8599. 
I think this is a really important feature.

**Comment by ThomasFOG** [comment]
As @squarebananas noted, it is possible to some extent to load assets with going through the content builder, but it is limited to the most basic assets like images/textures and sounds.

I have a different question for you, @SpanishCrusader.
What would be more accessible:
- being able to use the content builder GUI with a screen reader?
- or knowing that you can just edit the ```.mgcb``` file in a text editor to setup your content? It's basically the same semantic as using the GUI, expect that it's text-based.

I'm asking this so that we can figure out if it's a matter of adding features to the GUI, or documenting better the file format.

**Reply by MonophonicOne** [reply]
I think it is safe to assume the answer to your question is the first option. @SpanishCrusader did mention that he finds manually editing the Content file tedious, especially further in the development of a project. So knowing that it is an option does not change anything for them.

@SpanishCrusader , you might find https://github.com/Martenfur/Nopipeline to your liking. I have not used it myself, but it seems to offer the best of both worlds. You don't have to constantly edit the `Content.mgcb` but you still get your resources processed through the content builder pipeline. The resulting build should be identical regardless of whether one uses MGCB Editor, manually edited `.mgcb` file, or Nopipeline.

**Reply by SpanishCrusader** [reply]
Hello there! Sorry for the late reply, and tanks a lot for the interest.
Well as MonophonicOne mentioned, idealy the best option would be to be able to use the Content Builder GUI without any issues.
Since this is the primary method to load assests into the game, It is the primary option to recieve new features and updates.

Also I believe fixing the content builder UI might be easier than implementing alternative ways to load content.
I feel the issues are caused by the UI framework.
Some of the reading issues I encountered are quite familiar.

I'm not sure which UI framework is being used for the content builder, but in my experience there're several app frameworks which aren't quite compatible with screen readers.
I can name some of them like WPF, Electron, Qt... I had really bad experiences using apps built using those frameworks.

Recently I started working with UWP and WinUI3 because I'm working on an app to track my student's attendance record at the school.
The compatibility with screen readers is just perfect, I haven't ran into any issues while using their UI components in my app.

Alternatively, another option could be exposing the content builder's APi so maybe it would be posssible to manage assets from a script.
I believe libGDX does something like this. I think you could build your assets project from the code

I remember you had to do something like this

TextureParameters parameters = new TextureParameters();

parameters.setFormat("png");
parameters.setFilter(TextureParameters.Linear);
parameters.setPath("Game/Graphics/Terrain/Dungeon.png");

And then you would call the content builder's api like
ContentBuilder.build(parameters)

Then the API would generate the xnb at the specify location using the specified parameters.

But, since it might require some extra work and it might be difficult to maintain two different ways to load assets into the game, I think the best option is to make the ContentBuilder's UI more accessible.

Thanks you all for the given answers.
I'll definitely try the no pipeline addon, maybe it is exactly what I'm looking for.

Thanks a lot and Merry christmas!
